File: src/Tree/ConstructBinaryTreefromPreorderandPostorderTraversal/ConstructBinaryTreefromPreorderandPostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int[] pre_arr;
    int[] post_arr;
    Map<Integer, Integer> map = new HashMap<>();
    int index = 0;
    public TreeNode constructFromPrePost(int[] pre, int[] post) {
        //Build the table
        pre_arr = pre;
        post_arr = post;
        for(int i = 0; i < post.length; i++){
            map.put(post[i], i);
        }
        return dfs(pre.length);
    }
    private TreeNode dfs(int preIdx){
        if(index >= pre_arr.length) return null;
        int curVal = pre_arr[index];
        int curIdx = map.get(curVal);
        if(preIdx < curIdx) return null;
        index++;
        TreeNode root = new TreeNode(curVal);
        root.left = dfs(curIdx);
        root.right = dfs(curIdx);
        return root;
    }
}

File: src/Tree/ConstructBinaryTreefromPreorderandPostorderTraversal/ConstructBinaryTreefromPreorderandPostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        map = dict()
        for i in range(len(postorder)):
            map[postorder[i]] = i
            
        index = 0 # point to the index in preorder array
        
        def dfs(preIndex: int):
            nonlocal index
            if index == len(preorder):
                return None
            
            root_val = preorder[index]
            
            cur_postorder_index = map[root_val]
            
            if preIndex < cur_postorder_index:
                return None
            
            index += 1
            
            left = dfs(cur_postorder_index)
            right = dfs(cur_postorder_index)
            
            root = TreeNode(val=root_val, left=left, right=right)
            
            return root
            
            
        return dfs(len(postorder))
